pkg/segment/results/mresults/metricresults.go

```diff
@@ -702,14 +703,83 @@ func CalculateInterval(timerangeSeconds uint32) (uint32, error) {
 
 func (r *MetricsResult) aggregateFromAllTimeseries(aggregation structs.Aggregation) error {
 
+	var err error
 	switch aggregation.AggregatorFunction {
 	case segutils.Count:
 		r.computeAggCount(aggregation)
-	// Todo: Add TopK and BottomK
+	case segutils.TopK:
+		err = r.computeExtremesKElements(aggregation.FuncConstant, -1.0, true)
+	case segutils.BottomK:
+		err = r.computeExtremesKElements(aggregation.FuncConstant, 1.0, false)
 	default:
 		return fmt.Errorf("aggregateFromAllTimeseries: Unsupported aggregation: %v", aggregation)
 	}
 
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// The larger the priority, the earlier it will be popped out. Since we use the value as the priority, for `topk`, the larger the value, the more we want it to remain in the priority queue. Therefore, its priority should be smaller.
+// For bottomk, it's the opposite
+func (r *MetricsResult) computeExtremesKElements(funcConstant float64, factor float64, isTopK bool) error {
```

Comment:
I think `factor` and `isTopK` are acting as redundant pieces of information. Do you only need to have one of them as a parameter?

Reply:
I need `isTopK` to check these 2 conditions. And `factor` can help us to generate priority for each pq item without using if statement. What do you think?
```

	if isTopK && (item.Priority*factor) > runningDS.runningEntries[i].runningVal {
		heap.Push(pq, item)
		continue
	}
	if !isTopK && (item.Priority) < runningDS.runningEntries[i].runningVal {
		heap.Push(pq, item)
		continue
	}
```